dds/DCPS/security/SSL/SignedDocument.cpp

```diff
@@ -127,172 +108,67 @@ bool SignedDocument::load(const std::string& uri, DDS::Security::SecurityExcepti
-{
-  const std::string start_str("Content-Type: text/plain"), end_str("dds>");
-
-  get_original(dst);
-
-  const size_t found_begin = dst.find(start_str);
-
-  if (found_begin != std::string::npos) {
-    dst.erase(0, found_begin + start_str.length());
-
-    const char* t = " \t\n\r\f\v";
-
-    dst.erase(0, dst.find_first_not_of(t));
-  } else {
-    return false;
-  }
-
-  const size_t found_end = dst.find(end_str);
-
-  if (found_end != std::string::npos) {
-    dst.erase(found_end + end_str.length());
-  } else {
-    return false;
-  }
-
-  return true;
-}
-
-class verify_signature_impl
-{
-public:
-  verify_signature_impl(PKCS7* doc, const std::string& content)
-    : doc_(doc)
-    , content_(content)
-    , store_(X509_STORE_new())
-    , store_ctx_(X509_STORE_CTX_new())
-    , reader_(BIO_new(BIO_s_mem()))
-  {
-    if (!store_) {
-      OPENDDS_SSL_LOG_ERR("X509_STORE_new failed");
-    }
-    if (!store_ctx_) {
-      OPENDDS_SSL_LOG_ERR("X509_STORE_CTX_new failed");
-    }
-    if (!reader_) {
-      OPENDDS_SSL_LOG_ERR("BIO_new failed");
-    }
-  }
-
-  ~verify_signature_impl()
-  {
-    X509_STORE_CTX_free(store_ctx_);
-    X509_STORE_free(store_);
-    BIO_free(reader_);
-  }
-
-  int operator()(const Certificate& ca, unsigned long int flags = 0)
-  {
-    if (!doc_) return 1;
-    if (0 == content_.length()) return 1;
-
-    if (1 != X509_STORE_add_cert(store_, ca.x_)) {
+bool SignedDocument::verify(const Certificate& ca)
+{
+  content_.clear();
+  verified_ = false;
+
+  X509_STORE* store = X509_STORE_new();
+  if (store) {
+    if (X509_STORE_add_cert(store, ca.x509()) == 1) {
+      BIO* filebuf = BIO_new(BIO_s_mem());
+      if (filebuf) {
+        if (BIO_write(filebuf, original_.get_buffer(), original_.length()) > 0) {
+          BIO* bcont = 0;
+          PKCS7* doc = SMIME_read_PKCS7(filebuf, &bcont);
+          if (doc) {
+            BIO* content = BIO_new(BIO_s_mem());
+            if (content) {
+              if (PKCS7_verify(doc, 0, store, bcont, content, PKCS7_TEXT) == 1) {
+                char* p = 0;
+                long size = BIO_get_mem_data(content, &p);
+                if (size >= 0) {
+                  content_ = std::string(p, size);
+                  verified_ = true;
+                } else {
+                  OPENDDS_SSL_LOG_ERR("BIO_get_mem_data failed");
+                }
+              } else {
+                OPENDDS_SSL_LOG_ERR("PKCS7_verify failed");
+              }
+              BIO_free(content);
+            } else {
+              OPENDDS_SSL_LOG_ERR("BIO_new failed");
+            }
+            if (bcont) {
+              BIO_free(bcont);
+            }
+            PKCS7_free(doc);
+          } else {
+            OPENDDS_SSL_LOG_ERR("SMIME_read_PKCS7 failed");
+          }
+        } else {
+          OPENDDS_SSL_LOG_ERR("BIO_write failed");
+        }
+        BIO_free(filebuf);
+      } else {
+        OPENDDS_SSL_LOG_ERR("BIO_new failed");
```

Comment:
Nesting this deep gets hard to read and the error logging (the line right here) is far away from the line that caused it.

Using RAII for the `*_free` functions allows the method to exit when it needs to.  An example is the `hkdf()` in `CryptoBuiltinImpl.cpp`